File: backend/src/controllers/game.controller.ts
import { Request, Response } from 'express';
import { GameLogicService } from '../services/game-logic.service';
import { StateRepository } from '../repo/state-repo';
import { GameState, PendingAbilityChoice } from '../models/game-state';
import { EventSpawnerService } from '../services/event-spawner.service';
import { EventRepository } from '../repo/event-repo';
import { CompanionService } from '../services/companion.service';
import { BattleService } from '../services/battle.service';
import { DeckService } from '../services/deck.service';
import { Companion } from '../models/companion';
import { Card } from '../models/card';
import { LevelingService } from '../services/leveling.service';

export class GameController {
  private repository = new StateRepository();
  private gameLogic = new GameLogicService();
  private eventSpawner = new EventSpawnerService();
  private companionService = new CompanionService();
    private battleService = new BattleService();
  private deckService   = new DeckService();
  private eventRepo     = new EventRepository();
  private leveling      = new LevelingService();

  async getState(req: Request, res: Response) {
    const state = await this.repository.load();
    res.json(state);
  }

  async saveState(req: Request, res: Response) {
    const state = req.body as GameState;
    await this.repository.save(state);
    res.status(201).json({ saved: true });
  }

  async movePlayer(req: Request, res: Response) {
    const state = await this.repository.load();
    const nextNodeId = String(req.body.nextNodeId);
    const updated = this.gameLogic.movePlayer(state, nextNodeId);
    await this.repository.save(updated);
    res.json(updated);
  }

  async playCard(req: Request, res: Response) {
    const state = await this.repository.load();
    const cardId = String(req.body.cardId);
    const updated = this.gameLogic.playCard(state, cardId);
    await this.repository.save(updated);
    res.json(updated);
  }

    async resetGame(req: Request, res: Response) {
    const state = await this.gameLogic.createInitialState();
    await this.repository.save(state);
    res.status(201).json(state);
  }

  async getCompanions(req: Request, res: Response) {
    const companions = await this.companionService.getAll();
    res.json(companions);
  }

    async getEvents(req: Request, res: Response) {
    const specs = await this.eventSpawner.getSpecs();
    res.json(specs);
  }

  async validateEvent(req: Request, res: Response) {
    const body = req.body as { eventType: string; count: number };
    const result = await this.eventSpawner.validateCount(body.eventType, Number(body.count));
    res.json(result);
  }

  async getEventDefinitions(req: Request, res: Response) {
    const defs = await this.eventRepo.getAll();
    res.json(defs);
  }

  async battlePlayCard(req: Request, res: Response) {
    const state = await this.repository.load();
    const { cardId, companionId, targetIds } = req.body as {
      cardId: string;
      companionId: string;
      targetIds?: string[];
    };
    const updated = await this.battleService.playCard(
      state,
      String(cardId),
      String(companionId),
      targetIds
    );
    await this.repository.save(updated);
    res.json(updated);
  }

  async battleEndTurn(req: Request, res: Response) {
    const state = await this.repository.load();
    const updated = await this.battleService.endTurn(state);
    await this.repository.save(updated);
    res.json(updated);
  }

  async battleStart(req: Request, res: Response) {
    const state   = await this.repository.load();
    const updated = await this.battleService.startBattle(state);
    await this.repository.save(updated);
    res.json(updated);
  }

  async battleDrawCard(req: Request, res: Response) {
    const state    = await this.repository.load();
    const drawn    = this.deckService.drawCards(state, 1);
    const updated  = {
      ...drawn,
      battle: drawn.battle
        ? { ...drawn.battle, log: [...drawn.battle.log, 'Player drew 1 card'] }
        : drawn.battle,
      history: [...drawn.history, 'drew a card'],
    };
    await this.repository.save(updated);
    res.json(updated);
  }

    async finalizeCompanions(req: Request, res: Response) {
    const state = await this.repository.load();
    const { companions, baseCards } = req.body as {
      companions: Companion[];
      baseCards:  Card[];
    };
    const { deck, cards } = this.deckService.buildStartingDeck(baseCards, companions);
        // Stamp nextLevelExp on each companion so the frontend never needs the formula.
    const stampedCompanions = companions.map(c => this.leveling.withNextLevelExp(c));
    const updated = {
      ...state,
      player:     { ...state.player, deck, hand: [], discard: [] },
      cards,
      companions: stampedCompanions,
      history: [...state.history, 'Companions finalised — starting deck built'],
    };
    await this.repository.save(updated);
    res.json(updated);
  }

  async battleEnd(req: Request, res: Response) {
    const state = await this.repository.load();

    if (!state.battle) {
      res.status(400).json({ error: 'No battle state found.' });
      return;
    }
    if (state.battle.active) {
      res.status(400).json({ error: 'Battle is still active.' });
      return;
    }

    // Battle already ended (collectRewards set active: false).
    // Persist as-is and return so the frontend can proceed.
    await this.repository.save(state);
    res.json(state);
  }

  async chooseAbility(req: Request, res: Response) {
    const state = await this.repository.load();
    const { companionId, abilityId } = req.body as { companionId: string; abilityId: string };

    const pending = state.pendingAbilityChoices ?? [];
    const choice  = pending.find(c => c.companionId === companionId);
    if (!choice) {
      res.status(400).json({ error: 'No pending ability choice for this companion.' });
      return;
    }

    const ability = choice.options.find(a => a.id === abilityId);
    if (!ability) {
      res.status(400).json({ error: 'Invalid ability choice.' });
      return;
    }

    const updatedCompanions = state.companions.map(c =>
      c.id === companionId
        ? { ...c, specialAbilities: [...(c.specialAbilities ?? []), ability] }
        : c
    );

    const remainingChoices = pending.filter(c => c.companionId !== companionId);

    const updated: GameState = {
      ...state,
      companions: updatedCompanions,
      pendingAbilityChoices: remainingChoices,
      history: [...state.history, `${choice.companionName} learned ${ability.name}`],
    };

    await this.repository.save(updated);
    res.json(updated);
  }

  async battleClaimReward(req: Request, res: Response) {
    const state = await this.repository.load();
    const { companionId, cardId } = req.body as {
      companionId: string;
      cardId: string;
    };

        if (!state.battle) {
      res.status(400).json({ error: 'No battle state found.' });
      return;
    }

    if (state.battle.active) {
      res.status(400).json({ error: 'Cannot claim rewards while battle is still active.' });
      return;
    }

    const reward = state.battle.pendingCardRewards.find(
      r => r.companionId === companionId
    );
    if (!reward) {
      res.status(400).json({ error: `No pending reward for companion ${companionId}.` });
      return;
    }

    const chosenCard = reward.cardOptions.find(c => c.id === cardId);
    if (!chosenCard) {
      res.status(400).json({ error: `Card ${cardId} is not a valid reward option.` });
      return;
    }

    // 1. Add the chosen card to the master card catalogue.
    //    Reward card IDs are stamped with a unique suffix at reward-generation
    //    time, so a plain push is always safe — no deduplication needed.
    const updatedCards = [...state.cards, chosenCard];

    // 2. Add card id to the player's deck.
    const updatedDeck = {
      ...state.player.deck,
      cardIds: [...state.player.deck.cardIds, chosenCard.id],
    };

    // 3. Remove this reward from pendingCardRewards.
    const remainingRewards = state.battle.pendingCardRewards.filter(
      r => r.companionId !== companionId
    );

    const updated: GameState = {
      ...state,
      cards:  updatedCards,
      player: { ...state.player, deck: updatedDeck },
      battle: { ...state.battle, pendingCardRewards: remainingRewards },
      history: [
        ...state.history,
        `${companionId} claimed reward card: ${chosenCard.name}`,
      ],
    };

    await this.repository.save(updated);
    res.json(updated);
  }

}

File: backend-csharp/TheGameIsOn.API/Controllers/GameController.cs
using Microsoft.AspNetCore.Mvc;
using TheGameIsOn.API.Models;
using TheGameIsOn.API.Repositories;
using TheGameIsOn.API.Services;

namespace TheGameIsOn.API.Controllers;

// -------------------------------------------------------------------------
// Request DTOs
// -------------------------------------------------------------------------

public record MoveRequest(string NextNodeId);
public record PlayCardRequest(string CardId);
public record ValidateEventRequest(string EventType, int Count);
public record BattlePlayCardRequest(string CardId, string CompanionId, List<string>? TargetIds);
public record FinalizeCompanionsRequest(List<Companion> Companions);
public record ClaimRewardRequest(string CompanionId, string CardId);
public record ChooseAbilityRequest(string CompanionId, string AbilityId);

// -------------------------------------------------------------------------
// Controller
// -------------------------------------------------------------------------

[ApiController]
[Route("api/game")]
public class GameController : ControllerBase
{
    private readonly IStateRepository _repository;
    private readonly IGameLogicService _gameLogic;
    private readonly IEventSpawnerService _eventSpawner;
    private readonly ICompanionService _companionService;
    private readonly IBattleService _battleService;
    private readonly IDeckService _deckService;
    private readonly IEventRepository _eventRepo;
    private readonly ILevelingService _leveling;
    private readonly IBaseCardRepository _baseCardRepo;

    public GameController(
        IStateRepository repository,
        IGameLogicService gameLogic,
        IEventSpawnerService eventSpawner,
        ICompanionService companionService,
        IBattleService battleService,
        IDeckService deckService,
        IEventRepository eventRepo,
        ILevelingService leveling,
        IBaseCardRepository baseCardRepo)
    {
        _repository = repository;
        _gameLogic = gameLogic;
        _eventSpawner = eventSpawner;
        _companionService = companionService;
        _battleService = battleService;
        _deckService = deckService;
        _eventRepo = eventRepo;
        _leveling = leveling;
        _baseCardRepo = baseCardRepo;
    }

    // GET /api/game/state
    [HttpGet("state")]
    public async Task<IActionResult> GetState()
    {
        var state = await _repository.LoadAsync();
        return Ok(state);
    }

    // POST /api/game/state
    [HttpPost("state")]
    public async Task<IActionResult> SaveState([FromBody] GameState state)
    {
        await _repository.SaveAsync(state);
        return StatusCode(201, new { saved = true });
    }

    // POST /api/game/action/move
    [HttpPost("action/move")]
    public async Task<IActionResult> MovePlayer([FromBody] MoveRequest request)
    {
        var state = await _repository.LoadAsync();
        var updated = _gameLogic.MovePlayer(state, request.NextNodeId);
        await _repository.SaveAsync(updated);
        return Ok(updated);
    }

    // POST /api/game/action/play-card
    [HttpPost("action/play-card")]
    public async Task<IActionResult> PlayCard([FromBody] PlayCardRequest request)
    {
        var state = await _repository.LoadAsync();
        var updated = _gameLogic.PlayCard(state, request.CardId);
        await _repository.SaveAsync(updated);
        return Ok(updated);
    }

    // POST /api/game/action/new-run
    [HttpPost("action/new-run")]
    public async Task<IActionResult> ResetGame()
    {
        var state = await _gameLogic.CreateInitialStateAsync();
        await _repository.SaveAsync(state);
        return StatusCode(201, state);
    }

    // GET /api/game/action/companions
    [HttpGet("action/companions")]
    public async Task<IActionResult> GetCompanions()
    {
        var companions = await _companionService.GetAllAsync();
        return Ok(companions);
    }

    // GET /api/game/events
    [HttpGet("events")]
    public async Task<IActionResult> GetEvents()
    {
        var specs = await _eventSpawner.GetSpecsAsync();
        return Ok(specs);
    }

    // GET /api/game/events/definitions
    [HttpGet("events/definitions")]
    public async Task<IActionResult> GetEventDefinitions()
    {
        var defs = await _eventRepo.GetAllAsync();
        return Ok(defs);
    }

    // POST /api/game/events/validate
    [HttpPost("events/validate")]
    public async Task<IActionResult> ValidateEvent([FromBody] ValidateEventRequest request)
    {
        var (valid, reason) = await _eventSpawner.ValidateCountAsync(request.EventType, request.Count);
        return Ok(new { valid, reason });
    }

    // POST /api/game/action/battle/play-card
    [HttpPost("action/battle/play-card")]
    public async Task<IActionResult> BattlePlayCard([FromBody] BattlePlayCardRequest request)
    {
        var state = await _repository.LoadAsync();
        var updated = await _battleService.PlayCard(state, request.CardId, request.CompanionId, request.TargetIds);
        await _repository.SaveAsync(updated);
        return Ok(updated);
    }

    // POST /api/game/action/battle/end-turn
    [HttpPost("action/battle/end-turn")]
    public async Task<IActionResult> BattleEndTurn()
    {
        var state = await _repository.LoadAsync();
        var updated = await _battleService.EndTurn(state);
        await _repository.SaveAsync(updated);
        return Ok(updated);
    }

    // POST /api/game/action/battle/start
    [HttpPost("action/battle/start")]
    public async Task<IActionResult> BattleStart()
    {
        var state = await _repository.LoadAsync();
        var updated = await _battleService.StartBattle(state);
        await _repository.SaveAsync(updated);
        return Ok(updated);
    }

    // POST /api/game/action/battle/draw-card
    [HttpPost("action/battle/draw-card")]
    public async Task<IActionResult> BattleDrawCard()
    {
        var state = await _repository.LoadAsync();
        var drawn = _deckService.DrawCards(state, 1);

        var updated = new GameState
        {
            Player = drawn.Player,
            Graph = drawn.Graph,
            Cards = drawn.Cards,
            Companions = drawn.Companions,
            History = new List<string>(drawn.History) { "drew a card" },
            Battle = drawn.Battle is not null
                ? StatusEffectService.CloneBattle(drawn.Battle, b => b.Log.Add("Player drew 1 card"))
                : null,
            PendingAbilityChoices = drawn.PendingAbilityChoices
        };

        await _repository.SaveAsync(updated);
        return Ok(updated);
    }

    // POST /api/game/action/finalize-companions
    [HttpPost("action/finalize-companions")]
    public async Task<IActionResult> FinalizeCompanions([FromBody] FinalizeCompanionsRequest request)
    {
        var state = await _repository.LoadAsync();
        var baseCards = await _baseCardRepo.GetAllAsync();
        var (deck, cards) = _deckService.BuildStartingDeck(baseCards, request.Companions);
        var stampedCompanions = request.Companions.Select(c => _leveling.WithNextLevelExp(c)).ToList();

        var updated = new GameState
        {
            Player = state.Player with { Deck = deck, Hand = new(), Discard = new() },
            Graph = state.Graph,
            Cards = cards,
            Companions = stampedCompanions,
            History = new List<string>(state.History) { "Companions finalised — starting deck built" },
            Battle = state.Battle,
            PendingAbilityChoices = state.PendingAbilityChoices
        };

        await _repository.SaveAsync(updated);
        return Ok(updated);
    }

    // POST /api/game/action/battle/end
    [HttpPost("action/battle/end")]
    public async Task<IActionResult> BattleEnd()
    {
        var state = await _repository.LoadAsync();

        if (state.Battle is null)
            return BadRequest(new { error = "No battle state found." });
        if (state.Battle.Active)
            return BadRequest(new { error = "Battle is still active." });

        await _repository.SaveAsync(state);
        return Ok(state);
    }

    // POST /api/game/action/battle/claim-reward
    [HttpPost("action/battle/claim-reward")]
    public async Task<IActionResult> BattleClaimReward([FromBody] ClaimRewardRequest request)
    {
        var state = await _repository.LoadAsync();

        if (state.Battle is null)
            return BadRequest(new { error = "No battle state found." });
        if (state.Battle.Active)
            return BadRequest(new { error = "Cannot claim rewards while battle is still active." });

        var reward = state.Battle.PendingCardRewards.FirstOrDefault(r => r.CompanionId == request.CompanionId);
        if (reward is null)
            return BadRequest(new { error = $"No pending reward for companion {request.CompanionId}." });

        var chosenCard = reward.CardOptions.FirstOrDefault(c => c.Id == request.CardId);
        if (chosenCard is null)
            return BadRequest(new { error = $"Card {request.CardId} is not a valid reward option." });

        var updatedCards = new List<Card>(state.Cards) { chosenCard };
        var updatedDeck = state.Player.Deck with
        {
            CardIds = new List<string>(state.Player.Deck.CardIds) { chosenCard.Id }
        };
        var remainingRewards = state.Battle.PendingCardRewards
            .Where(r => r.CompanionId != request.CompanionId).ToList();

        var updated = new GameState
        {
            Player = state.Player with { Deck = updatedDeck },
            Graph = state.Graph,
            Cards = updatedCards,
            Companions = state.Companions,
            History = new List<string>(state.History)
            {
                $"{request.CompanionId} claimed reward card: {chosenCard.Name}"
            },
            Battle = StatusEffectService.CloneBattle(state.Battle, b => b.PendingCardRewards = remainingRewards),
            PendingAbilityChoices = state.PendingAbilityChoices
        };

        await _repository.SaveAsync(updated);
        return Ok(updated);
    }

    // POST /api/game/action/choose-ability
    [HttpPost("action/choose-ability")]
    public async Task<IActionResult> ChooseAbility([FromBody] ChooseAbilityRequest request)
    {
        var state = await _repository.LoadAsync();
        var pending = state.PendingAbilityChoices ?? new();
        var choice = pending.FirstOrDefault(c => c.CompanionId == request.CompanionId);

        if (choice is null)
            return BadRequest(new { error = "No pending ability choice for this companion." });

        var ability = choice.Options.FirstOrDefault(a => a.Id == request.AbilityId);
        if (ability is null)
            return BadRequest(new { error = "Invalid ability choice." });

        var updatedCompanions = state.Companions.Select(c =>
            c.Id == request.CompanionId
                ? CardEffectService.CloneCompanion(c, comp =>
                    comp.SpecialAbilities = new List<SpecialAbility>(comp.SpecialAbilities) { ability })
                : c
        ).ToList();

        var remainingChoices = pending.Where(c => c.CompanionId != request.CompanionId).ToList();

        var updated = new GameState
        {
            Player = state.Player,
            Graph = state.Graph,
            Cards = state.Cards,
            Companions = updatedCompanions,
            History = new List<string>(state.History) { $"{choice.CompanionName} learned {ability.Name}" },
            Battle = state.Battle,
            PendingAbilityChoices = remainingChoices
        };

        await _repository.SaveAsync(updated);
        return Ok(updated);
    }
}
